LibShifterBox 18 - Baertram -> PLEASE TEST too!

Baertram opened the pull request:
-Added function "errorText" for localized asserts and strip redundant…
… strings

-Changed asserts to use the function errorText
-Changed the possible/allowed events to be in a table and be dynamically added to the lib
-Added more callback events
-Added a new function "_fireCallback" to call the registered callbacks (to reduce the redundant code)
-Added new _assertFunction to the settings assert function
-Added new settings for the left and right list:
--Basically all parameters of the function ZO_ScrollList_AddDataType:
---rowOnMouseEnter: A function to call as the mouse enters the row of the list (e.g. show custom tooltip)
---rowOnMouseExit: A function to call as the mouse exits the row of the list (e.g. hide custom tooltip)
---rowOnMouseRightClick: A function to call as you right click on a row (e.g. show a context menu)
---rowSetupCallback: The setup callback function which should define the total row. Attention: It will call self:SetupRowEntry(rowControl, dataTabEnriched) automatically first! After that it will call your function here, and finally it will call  ZO_SortFilterList.SetupRow(self, rowControl, rowData)).
---rowDataTypeSelectSound: The sound (must be existing in global table SOUNDS) to play as a row with the datatype was selected
---rowResetControlCallback: A function to call as the control of the datatype was reset

--And in addition:
---rowSetupAdditionalDataCallback: A function to call AT THE START of the standard setupCallback function self:SetupRowEntry(rowControl, dataTabEnriched), which is able to change the entries of the data table (for the OnMouseEnter e.g. so that you got the data.tooltipText entry for the tooltip).

-Added new settings table entry "callbackRegister" where the key needs to be one of the available callbacks like lib.EVENT_LEFT_LIST_CREATED and the value a function for the callback. (This was needed to make the "list created" callback work as else the callback would only b registered AFTER the list were created already, and thus never fires. But it also maybe easier/better to find to add them all to the settings and let the for ..loop i the init function register them).

-Changed some funcitons to use less redundant Strings and code (like the assert settings functions)

Example settings could be e.g.:
local exampleLibShifterBoxCustomSettings = {
    leftList = {
        title = "Available sounds",
        callbackRegister = {
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_CREATED]            = function()
                d("LSB: Event left list created")
            end,
            [FCOChangeStuff.LSB.EVENT_RIGHT_LIST_CREATED]           = function()
                d("LSB: Event right list created")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_MOUSE_ENTER] = function()
                d("[FCOCS]LSB event left list row on mouse enter")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_MOUSE_EXIT]  = function()
                d("[FCOCS]LSB event left list row on mouse exit")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_DRAG_START]  = function()
                d("[FCOCS]LSB event left list row on drag start")
            end,
            [FCOChangeStuff.LSB.EVENT_RIGHT_LIST_ROW_ON_DRAG_END]   = function()
                d("[FCOCS]LSB event right list row on drag end")
            end,
        },
        rowOnMouseEnter = function(rowControl)
            local data = ZO_ScrollList_GetData(rowControl)
            d("LSB: OnMouseEnter: " ..tostring(data.tooltipText))
        end,
        rowOnMouseExit = function(rowControl) d("LSB: OnMouseExit")  end,
        rowOnMouseRightClick = function(rowControl, data) d("LSB: OnMouseRightClick") end,
        rowSetupCallback = function(rowControl, data)
            d("LSB: SetupCallback -> Calls self:SetupRowEntry, then this function, finally ZO_SortFilterList.SetupRow")
            data.tooltipText = "Hello world"
        end,
        rowDataTypeSelectSound = "ACTIVE_SKILL_RESPEC_MORPH_CHOSEN",
        rowResetControlCallback = function() d("LSB: ResetControlCallback")  end,
        rowSetupAdditionalDataCallback = function(rowControl, data)
            d("LSB: SetupAdditionalDataCallback")
                data.tooltipText = data.value
            return rowControl, data
        end,
    },
    rightList = {
        title = "Disabled sounds",
    }
}

Baertram committed:
-Added function "errorText" for localized asserts and strip redundant strings
-Changed asserts to use the function errorText
-Changed the possible/allowed events to be in a table and be dynamically added to the lib
-Added more callback events
-Added a new function "_fireCallback" to call the registered callbacks (to reduce the redundant code)
-Added new _assertFunction to the settings assert function
-Added new settings for the left and right list:
--Basically all parameters of the function ZO_ScrollList_AddDataType:
---rowOnMouseEnter: A function to call as the mouse enters the row of the list (e.g. show custom tooltip)
---rowOnMouseExit: A function to call as the mouse exits the row of the list (e.g. hide custom tooltip)
---rowOnMouseRightClick: A function to call as you right click on a row (e.g. show a context menu)
---rowSetupCallback: The setup callback function which should define the total row. Attention: You need to copy the function self:SetupRowEntry(rowControl, dataTabEnriched) in order to make the list work properly!!!, and afterwards enhance it where needed).
---rowDataTypeSelectSound: The sound (must be existing in global table SOUNDS) to play as a row with the datatype was selected
---rowResetControlCallback: A function to call as the control of the datatype was reset

--And in addition:
---rowSetupAdditionalDataCallback: A function to call AT THE START of the standard setupCallback function self:SetupRowEntry(rowControl, dataTabEnriched), which is able to change the entries of the data table (for the OnMouseEnter e.g. so that you got the data.tooltipText entry for the tooltip).

Baertram committed:
-Fixed function name to "_errorText"
-Changed indents

Baertram committed:
-Fixed _errorText call in line 13

Baertram committed:
-Changed the custom setupCallback function to call self:SetupRowEntry(rowControl, dataTabEnriched) internally first, and after that the custom function, and finally ZO_SortFilterList.SetupRow(self, rowControl, rowData)

Baertram committed:
-Added RegisterCallback to the settings menu in order to be able to specify a callback function for e.g. EVENT_LEFT_LIST_CREATED and EVENT_RIGHT_LIST_CREATED. Else the callback would be registered after the lists got created and would never fire :-)

Baertram committed:
-Added settings validation for the new registerCallback table
-Changed the validation functions to less redundant code

Baertram committed:
-Changed parameters of the ShifterBoxList:Initialize function: Added shifterBox from :New as else it is nil, as it get's added to the created control AFTER the init function
-Removed debug messages

Baertram commented:
Example settings could be e.g.:

``` 
local exampleLibShifterBoxCustomSettings = {
    leftList = {
        title = "Available sounds",
        callbackRegister = {
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_CREATED]            = function()
                d("LSB: Event left list created")
            end,
            [FCOChangeStuff.LSB.EVENT_RIGHT_LIST_CREATED]           = function()
                d("LSB: Event right list created")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_MOUSE_ENTER] = function()
                d("[FCOCS]LSB event left list row on mouse enter")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_MOUSE_EXIT]  = function()
                d("[FCOCS]LSB event left list row on mouse exit")
            end,
            [FCOChangeStuff.LSB.EVENT_LEFT_LIST_ROW_ON_DRAG_START]  = function()
                d("[FCOCS]LSB event left list row on drag start")
            end,
            [FCOChangeStuff.LSB.EVENT_RIGHT_LIST_ROW_ON_DRAG_END]   = function()
                d("[FCOCS]LSB event right list row on drag end")
            end,
        },
        rowOnMouseEnter = function(rowControl)
            local data = ZO_ScrollList_GetData(rowControl)
            d("LSB: OnMouseEnter: " ..tostring(data.tooltipText))
        end,
        rowOnMouseExit = function(rowControl) d("LSB: OnMouseExit")  end,
        rowOnMouseRightClick = function(rowControl, data) d("LSB: OnMouseRightClick") end,
        rowSetupCallback = function(rowControl, data)
            d("LSB: SetupCallback -> Calls self:SetupRowEntry, then this function, finally ZO_SortFilterList.SetupRow")
            data.tooltipText = "Hello world"
        end,
        rowDataTypeSelectSound = "ACTIVE_SKILL_RESPEC_MORPH_CHOSEN",
        rowResetControlCallback = function() d("LSB: ResetControlCallback")  end,
        rowSetupAdditionalDataCallback = function(rowControl, data)
            d("LSB: SetupAdditionalDataCallback")
                data.tooltipText = data.value
            return rowControl, data
        end,
    },
    rightList = {
        title = "Disabled sounds",
    }
}
```

Baertram committed:
-Added the 2nd return param "instance of the shifter box" to the function
--lib.GetControl
--to avoid an additional call to lib.GetShifterBox (where needed)

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -217,35 +257,49 @@ local function _applyCustomSettings(customSettings)
-            assert(type(customValue) == "number" and customValue > 0, string.format(LIB_IDENTIFIER.."_Error: Invalid %s parameter '%s' provided! Must be a numeric and positive.", parameterName, tostring(customValue)))
+            local assertionBool = ((typeText ~= "stringValue" and typeText ~= "sound") and type(customValue) == typeText) or false
+            if typeText == "number" and assertionBool == true then
+                assertionBool = customValue > 0
```
I'm wondering whether it's a good idea to do it like that.
Yes, there was no validation for just "number " (positive or negative) before, but it was clearly encapsulated in the "_assertPositiveNumber" function. Now the "_validateType" function would always only accept positive numbers when the typeText is set to "number" and it would not be possible to simply test for also negative number.

Baertram replied:
Yep, I'll change this to:
_validateType(customSettingsTbl, parameterName, settingsTbl, "number+")


``` 
-- validation functions
    local function _validateType(customSettingsTbl, parameterName, settingsTbl, typeText)
        local specialTypeTexts = {
            ["number+"]     = true,
            ["number-"]     = true,
            ["stringValue"] = true,
            ["sound"]       = true,
        }
        local customValue = customSettingsTbl[parameterName]
        if customValue ~= nil then
            local isSpecialTypeText = specialTypeTexts[typeText] or false
            local assertionBool = (not isSpecialTypeText and type(customValue) == typeText) or false
            if typeText == "number+" then
                assertionBool = customValue > 0
                typeText = typeText .. " and positive"
            elseif typeText == "number-" then
                assertionBool = customValue < 0
                typeText = typeText .. " and negative"
            elseif typeText == "stringValue" then
                assertionBool = (type(customValue) == "string" and (customValue == "value" or customValue == "key")) or false
                typeText = "either \'value\' or \'key\'"
            elseif typeText == "sound" then
                local sounds = SOUNDS
                assertionBool = (type(customValue) == "string" and sounds[customValue] ~= nil) or false
                typeText = "String and existing in global SOUNDS table"
            end
            assert(assertionBool == true, _errorText("Invalid %s parameter '%s' provided! Must be " .. tostring(typeText), parameterName, tostring(customValue)))
            settingsTbl[parameterName] = customValue
        end
    end
```

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -322,6 +393,10 @@ local function _removeEntriesFromList(list, keys)
     end
     if hasAtLeastOneRemoved then
+        -- then trigger the callback if present
+        _fireCallback(list.shifterBox, list, { [true] = lib.EVENT_LEFT_LIST_ENTRY_REMOVED, [false] = lib.EVENT_RIGHT_LIST_ENTRY_REMOVED  }, list.shifterBox.isLeftList,
```
Why not use "list.shifterBox.isLeftList" instead of [true] and [false] and then checking the same inside the "_fireCallback" function?
At least to me it seems a bit misleading to see true/false statements here for the two events but it actually does not control anything since it is only a reference to the "isLeftList" value.

Baertram replied:
I do not seem to understand you?

It controls the different event handlers used inside the function _fireCallback!
You pass in the table with the left (true) and right values (false) as "unique key".
Somehow the function needs the different event values (e.g. EVENT_LEFT_LIST_ENTRY_REMOVED and EVENT_RIGHT_LIST_ENTRY_REMOVED), so the table was the easiest way to achieve this.

The following 4th parameter list.shifterBox.isLeftList then will be the "key" of that table to use the event needed (true for left, or false for right event).

If you use the value list.shifterBox.isLeftList inside the table as key it would not be able to distinguish left and right? The key would be the same for both.

klingo replied:
I was rather thinking of something like this:
`_fireCallback(list.shifterBox, list, (list.shifterBox.isLeftList and lib.EVENT_LEFT_LIST_ENTRY_REMOVED or lib.EVENT_RIGHT_LIST_ENTRY_REMOVED), list.shifterBox.isLeftList,`
This way you can remove the check inside the "_fireCallback" function to figure out which event should be fired, because the function will only ever receive one event as a parameter. Makes it also more easy to read/understand from outside the function as you know exactly which event is passed on to the function, whereas in the current implementaiton this is not clear.

Baertram replied:
Good idea, implemented it.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -379,25 +454,57 @@ local function _addEntriesToList(list, entries, replace, otherList, categoryId)
+                    key=key,
+                    value=value,
+                    categoryId=categoryId,
+                    listRemovedFrom=(wasRemoved == 1 and list) or otherList,
```
Would it make sense to directly set the "list" or "otherList" in the "wasRemoved" parameter on line 466 instead of translating it to 1/2, and then translating 1/2 again into the lists? Could also be renamed to "listRemovedFrom" then.

Baertram replied:
Yep, good idea.
```
listRemovedFrom = (removeKey ~= nil and list) or otherList
...
if listRemovedFrom ~= nil then
                keysRemoved[key] = {
                    key=key,
                    value=value,
                    categoryId=categoryId,
                    listRemovedFrom=listRemovedFrom,
                }
            end
```

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -513,7 +657,13 @@ function ShifterBoxList:Initialize(control, shifterBoxSettings, isLeftList)
+                        local hasSameShifterBoxParent = _hasSameShifterBoxParent(self, sourceListControl)
+
+                        -- then trigger the callback if present
+                        _fireCallback(self.shifterBox, draggedOntoControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_DRAG_END, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_DRAG_END  }, isLeftList,
```
A) See previous comment from line 397.

B) Note to myself: _DRAG_END does not necessarily mean _DRAG_SUCCESSFUL (if dragging to an invalid, different shifterBox). I see that we need this event to e.g. clear anything related to dragging, but it might be misleading for consumers if thex expect that this event is only fired in case of a successful item-draging.

Baertram replied:
A) See above

B)
I guess the callback should fire after them, but without the "dragSuccessfull" value for now.
-> Currently the callback fires even for a "non valid" parent control, so the callback needs to fire inside "if hasSameShifterBoxParent then" !
And as you said, the info to the users needs to be "drag finished" is just as you released the mouse on a valid control. 

A "drag successfull" parameter could be added if we add some return values to the actual "items were moved after drag" functions.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -540,6 +690,14 @@ function ShifterBoxList:Initialize(control, shifterBoxSettings, isLeftList)
+        end
+    end
+    -- then trigger the callback if present
+    _fireCallback(shifterBox, control, { [true] = lib.EVENT_LEFT_LIST_CREATED, [false] = lib.EVENT_RIGHT_LIST_CREATED  }, isLeftList, shifterBox, isLeftList)
```
A) See previous comment from line 397.

B) See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -513,7 +657,13 @@ function ShifterBoxList:Initialize(control, shifterBoxSettings, isLeftList)
+
+                        -- then trigger the callback if present
+                        _fireCallback(self.shifterBox, draggedOntoControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_DRAG_END, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_DRAG_END  }, isLeftList,
+                                self.shifterBox, isLeftList, draggedOntoControl, mouseButton, dragData, hasSameShifterBoxParent)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -379,25 +454,57 @@ local function _addEntriesToList(list, entries, replace, otherList, categoryId)
         if hasAtLeastOneAdded then
+            -- then trigger the callback if present
+            _fireCallback(list.shifterBox, list, { [true] = lib.EVENT_LEFT_LIST_ENTRY_ADDED, [false] = lib.EVENT_RIGHT_LIST_ENTRY_ADDED  }, list.isLeftList,
+                    list.shifterBox, list.isLeftList, list, keysAdded)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -379,25 +454,57 @@ local function _addEntriesToList(list, entries, replace, otherList, categoryId)
         end
         if hasAtLeastOneAdded then
+            -- then trigger the callback if present
+            _fireCallback(list.shifterBox, list, { [true] = lib.EVENT_LEFT_LIST_ENTRY_ADDED, [false] = lib.EVENT_RIGHT_LIST_ENTRY_ADDED  }, list.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -322,6 +393,10 @@ local function _removeEntriesFromList(list, keys)
     if hasAtLeastOneRemoved then
+        -- then trigger the callback if present
+        _fireCallback(list.shifterBox, list, { [true] = lib.EVENT_LEFT_LIST_ENTRY_REMOVED, [false] = lib.EVENT_RIGHT_LIST_ENTRY_REMOVED  }, list.shifterBox.isLeftList,
+                list.shifterBox, list.shifterBox.isLeftList, list, keys)
```
Why do we need the "list.shifterBox", "list.shifterBox.isLeftList" and "list" parameters twice for calling "_fireCallback"? Can't you just use the value from the first (explicit) parameter?

Baertram replied:
-Fixed 1st param of _fireCallback (if no control is given teh shifterBox was passed to the callback function already)

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
             -- then trigger the callback if present
-            local callbackIdentifier = _getUniqueShifterBoxEventName(self.shifterBox, EVENT_ENTRY_HIGHLIGHTED)
-            CALLBACK_MANAGER:FireCallbacks(callbackIdentifier, control, self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
+            _fireCallback(self.shifterBox, control, { [true] = lib.EVENT_ENTRY_HIGHLIGHTED, [false] = lib.EVENT_ENTRY_HIGHLIGHTED  }, self.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
             -- then trigger the callback if present
-            local callbackIdentifier = _getUniqueShifterBoxEventName(self.shifterBox, EVENT_ENTRY_UNHIGHLIGHTED)
-            CALLBACK_MANAGER:FireCallbacks(callbackIdentifier, control, self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
+            _fireCallback(self.shifterBox, control, { [true] = lib.EVENT_ENTRY_UNHIGHLIGHTED, [false] = lib.EVENT_ENTRY_UNHIGHLIGHTED  }, self.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
+    doNotSetupRowNow = doNotSetupRowNow or false
+    local function onRowMouseEnter(p_rowControl)
+        -- then trigger the callback if present
+        _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_MOUSE_ENTER, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_MOUSE_ENTER  }, self.shifterBox.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
-        ZO_Tooltips_HideTextTooltip()
+    local function onRowMouseExit(p_rowControl)
+        -- then trigger the callback if present
+        _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_MOUSE_EXIT, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_MOUSE_EXIT  }, self.shifterBox.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
-            self:ToggleEntrySelection(data, rowControl, RESELECTING_DURING_REBUILD, false)
+    local function onRowMouseUp(p_rowControl, mouseButton, isInside, ctrlKey, altKey, shiftKey, commandKey)
+        -- then trigger the callback if present
+        _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_MOUSE_UP, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_MOUSE_UP  }, self.shifterBox.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
             lib.currentDragData  = currentDragData
+
+            -- then trigger the callback if present
+            _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_DRAG_START, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_DRAG_START  }, self.shifterBox.isLeftList,
```
See previous comment from line 397.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
-            local callbackIdentifier = _getUniqueShifterBoxEventName(self.shifterBox, EVENT_ENTRY_HIGHLIGHTED)
-            CALLBACK_MANAGER:FireCallbacks(callbackIdentifier, control, self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
+            _fireCallback(self.shifterBox, control, { [true] = lib.EVENT_ENTRY_HIGHLIGHTED, [false] = lib.EVENT_ENTRY_HIGHLIGHTED  }, self.isLeftList,
+                    self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
-            local callbackIdentifier = _getUniqueShifterBoxEventName(self.shifterBox, EVENT_ENTRY_UNHIGHLIGHTED)
-            CALLBACK_MANAGER:FireCallbacks(callbackIdentifier, control, self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
+            _fireCallback(self.shifterBox, control, { [true] = lib.EVENT_ENTRY_UNHIGHLIGHTED, [false] = lib.EVENT_ENTRY_UNHIGHLIGHTED  }, self.isLeftList,
+                    self.shifterBox, dataKey, data.value, data.categoryId, self.isLeftList)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
+    local function onRowMouseExit(p_rowControl)
+        -- then trigger the callback if present
+        _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_MOUSE_EXIT, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_MOUSE_EXIT  }, self.shifterBox.isLeftList,
+                self.shifterBox, self.shifterBox.isLeftList, p_rowControl, rowData)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
+
+            -- then trigger the callback if present
+            _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_DRAG_START, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_DRAG_START  }, self.shifterBox.isLeftList,
+                    self.shifterBox, self.shifterBox.isLeftList, p_rowControl, mouseButton, currentDragData)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -710,54 +868,86 @@ function ShifterBoxList:ToggleEntrySelection(data, control, reselectingDuringReb
+    local function onRowMouseEnter(p_rowControl)
+        -- then trigger the callback if present
+        _fireCallback(self.shifterBox, p_rowControl, { [true] = lib.EVENT_LEFT_LIST_ROW_ON_MOUSE_ENTER, [false] = lib.EVENT_RIGHT_LIST_ROW_ON_MOUSE_ENTER  }, self.shifterBox.isLeftList,
+                self.shifterBox, self.shifterBox.isLeftList, p_rowControl, rowData)
```
See previous comment from line 398.

klingo commented on LibShifterBox/LibShifterBox.txt:
```diff
@@ -6,10 +6,10 @@
-## Version: {VERSION_NUMBER}
-## AddOnVersion: {BUILD_NUMBER}
+## Version: 0.4.2
+## AddOnVersion: 18
```
Please re-add the placeholders for "Version" and "AddOnVersion", otherwise the build-script won't work properly anymore :)

Baertram committed:
-Changed code lines according to klingo's comments at the pull request

Baertram committed:
-Changed the _fireCallback function's 3rd parameter to a single eventId value
-Removed the _fireCallback function's 4th parameter "isLeftList"
-Added return values to some "move entry/entries" functions for the callback "DragEnd" parameter "was drag successfull"
-Checked and changed callback function parameters:
--removed duplicate values (e.g. controls already given as 1st param of the callback function -> see 2nd param of calling function _fireCallback)
--Corrected wrong isLeftList variables
--Made other parameters the same as existing EVENT* callback functions (where possible, e.g. pass the shifterBox as 1st param, or as 2nd if a control is given)

Baertram committed:
-Fixed 1st param of _fireCallback (if no control is given teh shifterBox was passed to the callback function already)

klingo commented on LibShifterBox/LibShifterBox.lua:
```diff
@@ -410,17 +542,23 @@ local function _addEntryToList(list, key, value, replace, otherList, categoryId)
 
 local function _moveEntriesToOtherList(sourceList, keys, destList, shifterBox)
+    local retVarLoop = false
+    local retVar = true
```
What is this variable (retVar) used for? From what I can see you initialize it with 'true' and at some point maybe set it to 'false', but then nothing happens with it? :)

Or should the returned parameter be retVarLoop? (As found in eg. toRightButtonClicked)

Baertram replied:
Whoops, the return variable should have been retVar instead of retVarLoop, like in toRightButtonClicked. Fixed it, thanks.

Baertram committed:
Update LibShifterBox.lua

Fixed return retVarLoop -> return retVar

Baertram committed:
Repaired EVENT_ENTRY_MOVED,readded isLeftList bool

Repaired the EVENT_ENTRY_MOVED parameters -> re-added isLeftList boolean as 4th parameter

Baertram committed:
Fixed EVENT_MOVED_ENTRY fromList->toList isLeft

Fixed EVENT_MOVED_ENTRY fromList->toList.isLeftList

Baertram committed:
LibShifterBox - Added drag&drop
